csrc/device_lower/pass/allocation.cpp

```diff
@@ -378,33 +390,96 @@ class AllocationInserter : public kir::ExprMutator {
     // solution, we should remove this after we have a better way to handle
     // synchronizations for cp.async.bulk.
     if (ir_utils::isCpAsyncBulkLoad(expr)) {
-      // create and allocate a memory barrier
-      TensorView* mbarrier = TensorViewBuilder()
-                                 .shape(std::vector<int64_t>{})
-                                 .dtype(DataType::UInt)
-                                 .contiguity(true)
-                                 .build();
-      mbarrier->setMemoryType(MemoryType::Shared);
-      auto mbarrier_init = IrBuilder::create<kir::MBarrierInit>(
-          mbarrier,
-          simplifyExpr(SimplifyingIrBuilder::maybeCastExpr(
-              DataType::UInt32,
-              lower_utils::getNumThreadsInTensorView(
-                  expr->output(0)->as<TensorView>()))));
-      auto sync_init = IrBuilder::create<kir::BlockSync>();
-      auto mbarrier_inval =
-          IrBuilder::create<kir::MBarrierInvalidate>(mbarrier);
-      auto sync_inval = IrBuilder::create<kir::BlockSync>();
-
-      kir::Allocate* mbarrier_alloc =
-          IrBuilder::create<kir::Allocate>(mbarrier, MemoryType::Shared);
-      Scope* expr_scope = scope_.empty() ? nullptr : scope_.back();
-      registerInsertBefore(expr, mbarrier_alloc, expr_scope);
-      registerInsertBefore(expr, mbarrier_init, expr_scope);
-      registerInsertBefore(expr, sync_init, expr_scope);
-      registerInsertAfter(expr, mbarrier_inval, expr_scope);
-      registerInsertAfter(expr, sync_inval, expr_scope);
-      GpuLower::current()->ldstMBarrierMap()[expr] = mbarrier;
+      if (circular_buffer_depth > 1) {
+        // Create and allocate a memory barrier. If this is a circular buffer,
+        // then allocate an array of mbarier objects. mbarrier::init and
+        // mbarrier::inval will be updated in circular buffering pass, but we
+        // add them here to handle shared memory correctly in alias memory pass.
+        TensorView* mbarrier =
+            TensorViewBuilder()
+                .shape(std::vector<int64_t>{circular_buffer_depth})
+                .dtype(DataType::UInt)
+                .contiguity(true)
+                .build();
+        mbarrier->setMemoryType(MemoryType::Shared);
+
+        // The wait condition for mbarrier is a single thread and the expected
```

Comment:
Instead of assuming there will be only a single thread issuing TMA load, should we check the parallelization of `expr->output(0)` to determine that value?

Reply:
The number of threads needs to a constant when we call `arrive_expected_tx` in the cuda kernel. The number of threads can be a runtime value, so defaulting to a single thread is the simplest thing.

Reply:
I'm using a regular `block_sync` for synchronizing the CTA rather than the `mbarrier` because of this compile-time constraint.

Reply:
I still don't understand. Could you elaborate more on:
> The number of threads needs to a constant when we call arrive_expected_tx in the cuda kernel.

Reply:
> The number of threads needs to a constant when we call arrive_expected_tx in the cuda kernel.

I was wrong about this ^^^.

I think your proposal is this:
```cpp
kir::MBarrierInit* mbarrier_init = IrBuilder::create<kir::MBarrierInit>(
    mbarrier,
    simplifyExpr(SimplifyingIrBuilder::maybeCastExpr(
        DataType::UInt32,
        lower_utils::getNumThreadsInTensorView(
            expr->output(0)->as<TensorView>()))));
```
I'm fine with this change.